### PR TITLE
Better attribute completion

### PR DESCRIPTION
**Pull request by vasily-kirichenko**
It fixes https://github.com/Microsoft/visualfsharp/issues/2041

At attribute application position:
* suggest namespaces, modules and types only
* remove "Attribute" suffix from attribute types

### CODE REVIEW

**Comment by KevinRansom** [comment]
Now this is nice.

File: src/fsharp/vs/ServiceDeclarations.fs [hunk 1]
```diff
@@ -1363,11 +1371,14 @@ type FSharpDeclarationListInfo(declarations: FSharpDeclarationListItem[]) =
                 | [] -> failwith "Unexpected empty bag"
                 | items -> 
                     let glyphMajor, glyphMinor = GlyphOfItem(denv,items.Head)
+                    let nm = if atAttributeApplication && IsAttribute infoReader items.Head then nm.[0..nm.Length-10] else nm
```
Comment (vladima) [inline src/fsharp/vs/ServiceDeclarations.fs]: nit: can you add a comment?

**Comment by vladima** [comment]
Looks great!

File: src/fsharp/vs/ServiceDeclarations.fs (hunk 1 above)
Comment (vasily-kirichenko) [inline src/fsharp/vs/ServiceDeclarations.fs]: Done.

**Comment by saul** [comment]
Thanks for doing this so quickly - fantastic 👍

**Comment by vasily-kirichenko** [comment]
@dotnet-bot test this please

**Comment by cartermp** [comment]
Great stuff.

**Comment by dsyme** [comment]
Yes, this is great

**Comment by KevinRansom** [comment]
@vasily-kirichenko 

some autocomplete test failures may be related:

````
1) Failed : Tests.LanguageService.AutoCompletion.UsingMSBuild.Attribute.WhenAttachedToNothingInNamespace.Bug70080
Couldn't find 'AttributeUsageAttribute' in completion list
at Salsa.VsOpsUtils.AssertCompListContains$cont@220(String membername, Tuple`4[] completions, Unit unitVar) in D:\j\workspace\release_ci_pa---3f142ccc\vsintegration\tests\Salsa\SalsaUtils.fs:line 223
at Salsa.VsOpsUtils.AssertCompListContains(Tuple`4[] completions, String membername) in D:\j\workspace\release_ci_pa---3f142ccc\vsintegration\tests\Salsa\SalsaUtils.fs:line 220
at Salsa.VsOpsUtils.AssertCompListContainsAll(Tuple`4[] completions, FSharpList`1 expectedCompletions) in D:\j\workspace\release_ci_pa---3f142ccc\vsintegration\tests\Salsa\SalsaUtils.fs:line 238
at <StartupCode$VisualFSharp-Unittests>.$Tests.LanguageService.Completion.AssertAutoCompleteContains@59-2.Invoke(Tuple`4[] completions) in D:\j\workspace\release_ci_pa---3f142ccc\vsintegration\tests\unittests\Tests.LanguageService.Completion.fs:line 60
at Tests.LanguageService.AutoCompletion.UsingMSBuild.DoWithAutoCompleteUsingExtraRefs(FSharpList`1 refs, Boolean coffeeBreak, SourceFileKind fileKind, BackgroundRequestReason reason, FSharpList`1 code, String marker, FSharpFunc`2 f) in D:\j\workspace\release_ci_pa---3f142ccc\vsintegration\tests\unittests\Tests.LanguageService.Completion.fs:line 50
at Tests.LanguageService.AutoCompletion.UsingMSBuild.DoWithAutoComplete(Boolean coffeeBreak, SourceFileKind fileKind, BackgroundRequestReason reason, FSharpList`1 code, String marker, FSharpFunc`2 f) in D:\j\workspace\release_ci_pa---3f142ccc\vsintegration\tests\unittests\Tests.LanguageService.Completion.fs:line 54
at <StartupCode$VisualFSharp-Unittests>.$Tests.LanguageService.Completion.AssertAutoCompleteContains@58-1.Invoke(FSharpList`1 code, String marker, FSharpList`1 should, FSharpList`1 shouldnot) in D:\j\workspace\release_ci_pa---3f142ccc\vsintegration\tests\unittests\Tests.LanguageService.Completion.fs:line 58
at Tests.LanguageService.AutoCompletion.UsingMSBuild.AutoCompleteBug70080HelperHelper(String programText, FSharpList`1 shouldContain, FSharpList`1 shouldNotContain) in D:\j\workspace\release_ci_pa---3f142ccc\vsintegration\tests\unittests\Tests.LanguageService.Completion.fs:line 167

2) Failed : Tests.LanguageService.AutoCompletion.UsingMSBuild.LongIdent.AsAttribute
Couldn't find 'ObsoleteAttribute' in completion list
at Salsa.VsOpsUtils.AssertCompListContains$cont@220(String membername, Tuple`4[] completions, Unit unitVar) in D:\j\workspace\release_ci_pa---3f142ccc\vsintegration\tests\Salsa\SalsaUtils.fs:line 223
at Salsa.VsOpsUtils.AssertCompListContains(Tuple`4[] completions, String membername) in D:\j\workspace\release_ci_pa---3f142ccc\vsintegration\tests\Salsa\SalsaUtils.fs:line 220
at Salsa.VsOpsUtils.AssertCompListContainsAll(Tuple`4[] completions, FSharpList`1 expectedCompletions) in D:\j\workspace\release_ci_pa---3f142ccc\vsintegration\tests\Salsa\SalsaUtils.fs:line 238
at Salsa.VsOpsUtils.AssertCompListContainsAll(Tuple`4[] completions, FSharpList`1 expectedCompletions) in D:\j\workspace\release_ci_pa---3f142ccc\vsintegration\tests\Salsa\SalsaUtils.fs:line 239

3) Failed : Tests.LanguageService.AutoCompletion.UsingMSBuild.LongIdent.PInvoke.AsAttribute
Couldn't find 'SomeAttrib' in completion list
at Salsa.VsOpsUtils.AssertCompListContains$cont@220(String membername, Tuple`4[] completions, Unit unitVar) in D:\j\workspace\release_ci_pa---3f142ccc\vsintegration\tests\Salsa\SalsaUtils.fs:line 223
at Salsa.VsOpsUtils.AssertCompListContains(Tuple`4[] completions, String membername) in D:\j\workspace\release_ci_pa---3f142ccc\vsintegration\tests\Salsa\SalsaUtils.fs:line 220
at Salsa.VsOpsUtils.AssertCompListContainsAll(Tuple`4[] completions, FSharpList`1 expectedCompletions) in D:\j\workspace\release_ci_pa---3f142ccc\vsintegration\tests\Salsa\SalsaUtils.fs:line 238

4) Failed : Tests.LanguageService.AutoCompletion.UsingProjectSystem.Attribute.WhenAttachedToNothingInNamespace.Bug70080
Couldn't find 'AttributeUsageAttribute' in completion list
at Salsa.VsOpsUtils.AssertCompListContains$cont@220(String membername, Tuple`4[] completions, Unit unitVar) in D:\j\workspace\release_ci_pa---3f142ccc\vsintegration\tests\Salsa\SalsaUtils.fs:line 223
at Salsa.VsOpsUtils.AssertCompListContains(Tuple`4[] completions, String membername) in D:\j\workspace\release_ci_pa---3f142ccc\vsintegration\tests\Salsa\SalsaUtils.fs:line 220
at Salsa.VsOpsUtils.AssertCompListContainsAll(Tuple`4[] completions, FSharpList`1 expectedCompletions) in D:\j\workspace\release_ci_pa---3f142ccc\vsintegration\tests\Salsa\SalsaUtils.fs:line 238
at <StartupCode$VisualFSharp-Unittests>.$Tests.LanguageService.Completion.AssertAutoCompleteContains@59-2.Invoke(Tuple`4[] completions) in D:\j\workspace\release_ci_pa---3f142ccc\vsintegration\tests\unittests\Tests.LanguageService.Completion.fs:line 60
at Tests.LanguageService.AutoCompletion.UsingMSBuild.DoWithAutoCompleteUsingExtraRefs(FSharpList`1 refs, Boolean coffeeBreak, SourceFileKind fileKind, BackgroundRequestReason reason, FSharpList`1 code, String marker, FSharpFunc`2 f) in D:\j\workspace\release_ci_pa---3f142ccc\vsintegration\tests\unittests\Tests.LanguageService.Completion.fs:line 50
at Tests.LanguageService.AutoCompletion.UsingMSBuild.DoWithAutoComplete(Boolean coffeeBreak, SourceFileKind fileKind, BackgroundRequestReason reason, FSharpList`1 code, String marker, FSharpFunc`2 f) in D:\j\workspace\release_ci_pa---3f142ccc\vsintegration\tests\unittests\Tests.LanguageService.Completion.fs:line 54
at <StartupCode$VisualFSharp-Unittests>.$Tests.LanguageService.Completion.AssertAutoCompleteContains@58-1.Invoke(FSharpList`1 code, String marker, FSharpList`1 should, FSharpList`1 shouldnot) in D:\j\workspace\release_ci_pa---3f142ccc\vsintegration\tests\unittests\Tests.LanguageService.Completion.fs:line 58
at Tests.LanguageService.AutoCompletion.UsingMSBuild.AutoCompleteBug70080HelperHelper(String programText, FSharpList`1 shouldContain, FSharpList`1 shouldNotContain) in D:\j\workspace\release_ci_pa---3f142ccc\vsintegration\tests\unittests\Tests.LanguageService.Completion.fs:line 167

5) Failed : Tests.LanguageService.AutoCompletion.UsingProjectSystem.LongIdent.AsAttribute
Couldn't find 'ObsoleteAttribute' in completion list
at Salsa.VsOpsUtils.AssertCompListContains$cont@220(String membername, Tuple`4[] completions, Unit unitVar) in D:\j\workspace\release_ci_pa---3f142ccc\vsintegration\tests\Salsa\SalsaUtils.fs:line 223
at Salsa.VsOpsUtils.AssertCompListContains(Tuple`4[] completions, String membername) in D:\j\workspace\release_ci_pa---3f142ccc\vsintegration\tests\Salsa\SalsaUtils.fs:line 220
at Salsa.VsOpsUtils.AssertCompListContainsAll(Tuple`4[] completions, FSharpList`1 expectedCompletions) in D:\j\workspace\release_ci_pa---3f142ccc\vsintegration\tests\Salsa\SalsaUtils.fs:line 238
at Salsa.VsOpsUtils.AssertCompListContainsAll(Tuple`4[] completions, FSharpList`1 expectedCompletions) in D:\j\workspace\release_ci_pa---3f142ccc\vsintegration\tests\Salsa\SalsaUtils.fs:line 239

6) Failed : Tests.LanguageService.AutoCompletion.UsingProjectSystem.LongIdent.PInvoke.AsAttribute
Couldn't find 'SomeAttrib' in completion list
at Salsa.VsOpsUtils.AssertCompListContains$cont@220(String membername, Tuple`4[] completions, Unit unitVar) in D:\j\workspace\release_ci_pa---3f142ccc\vsintegration\tests\Salsa\SalsaUtils.fs:line 223
at Salsa.VsOpsUtils.AssertCompListContains(Tuple`4[] completions, String membername) in D:\j\workspace\release_ci_pa---3f142ccc\vsintegration\tests\Salsa\SalsaUtils.fs:line 220
at Salsa.VsOpsUtils.AssertCompListContainsAll(Tuple`4[] completions, FSharpList`1 expectedCompletions) in D:\j\workspace\release_ci_pa---3f142ccc\vsintegration\tests\Salsa\SalsaUtils.fs:line 238

````

**Comment by vasily-kirichenko** [comment]
Different attribute completiion tests fails from launch to launch. I think it's because completion list caching:
 
* a test executes completion at "no attribute application" point => "AnAttribute" completion is cached
* another test executes at "attribute application" point => the caches "AnAttribute" is returned instead of "An"

This behavior is also reproduced in the editor. We should clean the cache if we move from attr appl point to non such a point.

**Comment by vasily-kirichenko** [comment]
Where can I find why the tests failed? I examined all the files in this list:

image

and found nothing labeled as "error" or "failure".

**Comment by vasily-kirichenko** [comment]
@forki Thanks, but I didn't find any failures in the console output, the only thing that has helped is searching for "failure" in the text that's available by "View as plain text" link.

**Comment by vasily-kirichenko** [comment]
Hm. I've no idea why the tests do not pass now. The sample from one of them works in editor:

image

**Comment by vasily-kirichenko** [comment]
@dotnet-bot test this please

**Comment by vasily-kirichenko** [comment]
I think it's finally ready.

**Comment by KevinRansom** [comment]
Thanks for taking care of this.

Kevin

**Pull request merged**